File: scripts/refactor_backend.mjs
import fs from 'fs';
import path from 'path';

const BACKEND_DIR = path.resolve('c:/Users/murat/golabs-web/apps/backend');
const APP_DIR = path.join(BACKEND_DIR, 'app');
const CORE_DIR = path.join(APP_DIR, 'core');

function removeIfExists(p) {
    if (fs.existsSync(p)) {
        const stat = fs.statSync(p);
        if (stat.isDirectory()) {
            fs.rmSync(p, { recursive: true, force: true });
            console.log(`Removed directory: ${p}`);
        } else {
            fs.rmSync(p, { force: true });
            console.log(`Removed file: ${p}`);
        }
    }
}

// 1. Remove empty/duplicate folders and dummy files
removeIfExists(path.join(CORE_DIR, 'config'));
removeIfExists(path.join(CORE_DIR, 'exceptions'));
removeIfExists(path.join(CORE_DIR, 'whatsapp_service.py'));

// 2. Re-create / move services
removeIfExists(path.join(APP_DIR, 'services'));

const coreServicesDir = path.join(CORE_DIR, 'services');
const appServicesDir = path.join(APP_DIR, 'services');
if (fs.existsSync(coreServicesDir)) {
    fs.renameSync(coreServicesDir, appServicesDir);
    console.log("Moved app/core/services to app/services");
} else {
    if (!fs.existsSync(appServicesDir)) fs.mkdirSync(appServicesDir);
}

// 3. Create app/services/email
const emailDir = path.join(appServicesDir, 'email');
if (!fs.existsSync(emailDir)) {
    fs.mkdirSync(emailDir, { recursive: true });
    fs.writeFileSync(path.join(emailDir, '__init__.py'), '');
}

// Move email files
['emailing.py', 'email_templates.py'].forEach(f => {
    const src = path.join(CORE_DIR, f);
    if (fs.existsSync(src)) {
        fs.renameSync(src, path.join(emailDir, f));
        console.log(`Moved ${f} to app/services/email/`);
    }
});

// Move other standalone services
['notification_service.py', 'storage.py', 'upload_validator.py'].forEach(f => {
    const src = path.join(CORE_DIR, f);
    if (fs.existsSync(src)) {
        fs.renameSync(src, path.join(appServicesDir, f));
        console.log(`Moved ${f} to app/services/`);
    }
});

// 4. Move app/core/workers to app/workers
const coreWorkersDir = path.join(CORE_DIR, 'workers');
const appWorkersDir = path.join(APP_DIR, 'workers');
if (fs.existsSync(coreWorkersDir)) {
    fs.renameSync(coreWorkersDir, appWorkersDir);
    console.log("Moved app/core/workers to app/workers");
}

// 5. Update Imports
const REPLACEMENTS = [
    { pattern: /\bapp\.core\.services\b/g, repl: 'app.services' },
    { pattern: /\bapp\.core\.notification_service\b/g, repl: 'app.services.notification_service' },
    { pattern: /\bapp\.core\.storage\b/g, repl: 'app.services.storage' },
    { pattern: /\bapp\.core\.upload_validator\b/g, repl: 'app.services.upload_validator' },
    { pattern: /\bapp\.core\.emailing\b/g, repl: 'app.services.email.emailing' },
    { pattern: /\bapp\.core\.email_templates\b/g, repl: 'app.services.email.email_templates' },
    { pattern: /\bapp\.core\.workers\b/g, repl: 'app.workers' },
];

function processFile(filepath) {
    let content = fs.readFileSync(filepath, 'utf8');
    let originalContent = content;
    
    for (const { pattern, repl } of REPLACEMENTS) {
        content = content.replace(pattern, repl);
    }
    
    if (content !== originalContent) {
        fs.writeFileSync(filepath, content, 'utf8');
        console.log(`Updated imports in: ${filepath}`);
    }
}

function scanDir(dir) {
    if (!fs.existsSync(dir)) return;
    const files = fs.readdirSync(dir);
    for (const file of files) {
        const fullPath = path.join(dir, file);
        if (fullPath.includes('venv') || fullPath.includes('.pytest_cache') || fullPath.includes('__pycache__') || fullPath.includes('alembic')) continue;
        
        const stat = fs.statSync(fullPath);
        if (stat.isDirectory()) {
            scanDir(fullPath);
        } else if (file.endsWith('.py')) {
            processFile(fullPath);
        }
    }
}

scanDir(BACKEND_DIR);
console.log("Backend refactoring complete.");

File: scripts/refactor_backend.py
import os
import shutil
import re

BACKEND_DIR = os.path.abspath('c:/Users/murat/golabs-web/apps/backend')
APP_DIR = os.path.join(BACKEND_DIR, 'app')
CORE_DIR = os.path.join(APP_DIR, 'core')

def remove_if_exists(path):
    if os.path.isdir(path):
        shutil.rmtree(path)
        print(f"Removed directory: {path}")
    elif os.path.isfile(path):
        os.remove(path)
        print(f"Removed file: {path}")

# 1. Remove empty/duplicate folders and dummy files
remove_if_exists(os.path.join(CORE_DIR, 'config'))
remove_if_exists(os.path.join(CORE_DIR, 'exceptions'))
remove_if_exists(os.path.join(CORE_DIR, 'whatsapp_service.py'))

# 2. Re-create / move services
# First, remove the dummy app/services folder
remove_if_exists(os.path.join(APP_DIR, 'services'))

# Move app/core/services to app/services
core_services_dir = os.path.join(CORE_DIR, 'services')
app_services_dir = os.path.join(APP_DIR, 'services')
if os.path.exists(core_services_dir):
    shutil.move(core_services_dir, app_services_dir)
    print("Moved app/core/services to app/services")

# 3. Create app/services/email
email_dir = os.path.join(app_services_dir, 'email')
os.makedirs(email_dir, exist_ok=True)
with open(os.path.join(email_dir, '__init__.py'), 'w') as f:
    pass

# Move email files
for f in ['emailing.py', 'email_templates.py']:
    src = os.path.join(CORE_DIR, f)
    if os.path.exists(src):
        shutil.move(src, os.path.join(email_dir, f))
        print(f"Moved {f} to app/services/email/")

# Move other standalone services
for f in ['notification_service.py', 'storage.py', 'upload_validator.py']:
    src = os.path.join(CORE_DIR, f)
    if os.path.exists(src):
        shutil.move(src, os.path.join(app_services_dir, f))
        print(f"Moved {f} to app/services/")

# 4. Move app/core/workers to app/workers
core_workers_dir = os.path.join(CORE_DIR, 'workers')
app_workers_dir = os.path.join(APP_DIR, 'workers')
if os.path.exists(core_workers_dir):
    shutil.move(core_workers_dir, app_workers_dir)
    print("Moved app/core/workers to app/workers")

# 5. Update Imports
REPLACEMENTS = [
    (r'\bapp\.core\.services\b', 'app.services'),
    (r'\bapp\.core\.notification_service\b', 'app.services.notification_service'),
    (r'\bapp\.core\.storage\b', 'app.services.storage'),
    (r'\bapp\.core\.upload_validator\b', 'app.services.upload_validator'),
    (r'\bapp\.core\.emailing\b', 'app.services.email.emailing'),
    (r'\bapp\.core\.email_templates\b', 'app.services.email.email_templates'),
    (r'\bapp\.core\.workers\b', 'app.workers'),
]

def process_file(filepath):
    with open(filepath, 'r', encoding='utf-8') as f:
        content = f.read()

    original_content = content
    for pattern, repl in REPLACEMENTS:
        content = re.sub(pattern, repl, content)

    if content != original_content:
        with open(filepath, 'w', encoding='utf-8') as f:
            f.write(content)
        print(f"Updated imports in: {filepath}")

for root, _, files in os.walk(BACKEND_DIR):
    # skip venv, .pytest_cache, __pycache__, alembic
    if 'venv' in root or '.pytest_cache' in root or '__pycache__' in root or 'alembic' in root:
        continue
    for file in files:
        if file.endswith('.py'):
            process_file(os.path.join(root, file))

print("Backend refactoring complete.")
